### ISSUE TITLE
catch-all type for decorator value

### ISSUE THREAD

**Issue by trusktr** [microsoft/TypeScript#56426]
### 🔍 Search Terms

github typescript decorator value type

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

> This is a request to add a new utility type because TypeScript added one for decorator contexts but not for values

The `DecoratorContext` type is nice for catching all context types:

```js
function decorator(value: unknown, context: DecoratorContext) { // optionally use a more specific type.
  // ...
}
```

But for `value`, there is not a good catch-all type.

### 📃 Motivating Example

If we have a context catch all type, why not also have a value catch all type, f.e. something like :

```ts
export type DecoratedValue = Constructor | Function | ClassAccessorDecoratorTarget<object, unknown> | undefined
```

### 💻 Use Cases

1. What do you want to use this for? decorators whose code handles any value type
3. What shortcomings exist with current approaches? we have to make the catchall for value ourselves, but we don't have to do that for context
4. What workarounds are you using in the meantime? I make my own catchall (not that big a deal, but neither would making `DecoratorContext` myself be

**Comment by MartinJohns** [comment]
> [x] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types

But technically it is.

**Comment by trusktr** [comment]
> But technically it is.

Why at it, could it be helpful to introduce things like the following?

```ts
function isMemberDecorator(o: DecoratorContext): o is ClassMemberDecorator
```

because:

<img width="728" alt="Screenshot 2023-11-15 at 9 39 55 PM" src="https://github.com/microsoft/TypeScript/assets/297678/8b535f4b-1481-4044-97a5-cbae89ef104d">

And we have to do this:

<img width="734" alt="Screenshot 2023-11-15 at 9 40 17 PM" src="https://github.com/microsoft/TypeScript/assets/297678/bc3d1c25-9724-4ab2-b92a-2bbe04d107ab">

Decorator return types are tricky. There appear to not be any built-in types we can use. For example, I was hoping for something like `FieldInitializer`:

```ts
function fieldDeco(value: unknown, context: ClassFieldDecoratorContext): FieldInitializer | void
```

Etc for other return types. And also a catch all:

```ts
type DecoratorReturn = FieldInitializer | etc
```

**Comment by fatcerberus** [comment]
fwiw, I don't think `DecoratorContext` counts as a "utility type" in the usual sense; that point in the checklist generally refers to requests for types in the vein of `Partial` or `Omit` where the purpose is to manipulate other types (i.e. utility type = type utility).  IMO what's asked for here is not a utility type any more than, say, `ArrayBufferLike` or `BufferSource`.

Of course it's an open question whether the TS maintainers will want to add new named types that can easily be expressed directly as unions.  But that decision will likely be more pragmatic than the typical blanket prohibition against utility types.

**Comment by fatcerberus** [comment]
> Oh wait, does TypeScript even ship any type guards?

No, and it's unlikely ever to.  Outside of grandfathered features like `enum`, TS doesn't do runtime stuff.  Hence this point in the checklist

> This isn't a runtime feature (e.g. **library functionality**, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
